packages/arb-avm-cpp/data_storage/src/arbcore.cpp

```diff
@@ -1173,69 +1179,64 @@ rocksdb::Status ArbCore::advanceExecutionCursorImpl(
-            if (remaining_gas > 0) {
-                MachineExecutionConfig execConfig;
-                execConfig.max_gas = total_gas_used;
-                execConfig.go_over_gas = go_over_gas;
-                execConfig.inbox_messages =
-                    std::move(get_messages_result.data.second);
+            MachineExecutionConfig execConfig;
+            execConfig.max_gas = total_gas_used;
+            execConfig.go_over_gas = go_over_gas;
+
+            {
+                ReadSnapshotTransaction tx(data_storage);
+
+                auto& mach =
+                    resolveExecutionCursorMachine(tx, execution_cursor, cache);
 
                 // Resolve staged message if possible.
                 // If placeholder message not found, machine will just be
                 // blocked
-                auto resolve_status = std::visit(
-                    [&](auto& mach) {
-                        return resolveStagedMessage(
-                            tx, resolveExecutionVariant(mach));
-                    },
-                    execution_cursor.machine);
+                auto resolve_status =
+                    resolveStagedMessage(tx, mach->machine_state);
                 if (!resolve_status.IsNotFound() && !resolve_status.ok()) {
                     core_error_string = "error resolving staged message";
                     machine_error = true;
                     std::cerr << "ArbCore error: " << core_error_string << ": "
                               << resolve_status.ToString() << "\n";
                     return resolve_status;
                 }
-                auto& mach =
-                    resolveExecutionCursorMachine(tx, execution_cursor, cache);
-                mach->machine_state.context = AssertionContext(execConfig);
-                auto assertion = mach->run();
-                if (assertion.gasCount == 0) {
-                    // Nothing was executed
+
+                if (go_over_gas && execution_cursor.getOutput().arb_gas_used >=
+                                       total_gas_used) {
+                    break;
+                } else if (!go_over_gas &&
+                           mach->machine_state.nextGasCost() > total_gas_used) {
                     break;
                 }
 
-                if (total_gas_used <=
-                    execution_cursor.getOutput().arb_gas_used) {
-                    // Gas reached
+                auto get_messages_result = executionCursorGetMessages(
+                    tx, execution_cursor, message_group_size);
+                if (!get_messages_result.status.ok()) {
+                    std::cout << "Error getting messages for execution cursor"
+                              << std::endl;
+                    return get_messages_result.status;
+                }
+                if (!get_messages_result.data.first) {
+                    // Reorg occurred, need to recreate machine
+                    handle_reorg = true;
```

Comment:
I don't know if it would print too many logs, but at least for initial debugging it might be useful to have a debug message here to tell us the message/gas that it reorged at?

Reply:
Agreed. Added a `cout`